### PR TITLE
create simplify_dyn_ops pass and add variable slice optimization

### PR DESCRIPTION
**Pull request by CharlieL7**
* New compiler pass that simplifies dynamic shapes related operators to their static versions if possible
* Will normally be used after a `split_single_dyn_dim` pass

### CODE REVIEW

File: test/split_single_dyn_dim_test.cpp [hunk 1]
```diff
@@ -157,64 +157,4 @@ TEST_CASE(multiple_outputs)
     EXPECT(p0 == p1);
 }
 
-TEST_CASE(broadcast_match)
```
Comment (TedThemistokleous) [inline test/split_single_dyn_dim_test.cpp:160]: If i'm not mistaken your new pass should handle this case then and remove the need for this test?

File: src/split_single_dyn_dim.cpp [hunk 2]
```diff
@@ -68,37 +68,6 @@ has_one_dyn_dim(const std::unordered_map<std::string, shape>& param_shapes)
                                     dds_it->max};
 }
 
-namespace {
-struct find_static_2in_broadcasts
```
Comment (TedThemistokleous) [inline src/split_single_dyn_dim.cpp:72]: Same as my other comment below in terms of not needing this matcher anymore like we don't need the testcase

**Review by TedThemistokleous — COMMENTED** [review]
ORT failure seems to be unrelated to your changes right now. Overall just answer some Q's I have

File: src/split_single_dyn_dim.cpp (hunk 2 above)
Comment (CharlieL7) [reply src/split_single_dyn_dim.cpp:72]: I moved these finders to a new pass `simplify_dynamic_ops`

File: test/split_single_dyn_dim_test.cpp (hunk 1 above)
Comment (CharlieL7) [reply test/split_single_dyn_dim_test.cpp:160]: Yep

**Review by TedThemistokleous — APPROVED** [review]

File: src/simplify_dyn_ops.cpp [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+                                    match::arg(2)(match::is_constant()));
+    }
+
+    void apply(module& m, const match::matcher_result& mr) const
+    {
+        auto ins            = mr.result;
+        auto inputs         = ins->inputs();
+        argument starts_arg = inputs.at(1)->eval();
+        argument ends_arg   = inputs.at(2)->eval();
+        if(not(starts_arg.empty() and ends_arg.empty()))
```
Comment (umangyadav) [inline src/simplify_dyn_ops.cpp:81]:
```suggestion
        if(not starts_arg.empty() and not ends_arg.empty())
```

File: src/simplify_dyn_ops.cpp [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    }
+
+    void apply(module& m, const match::matcher_result& mr) const
+    {
+        auto ins            = mr.result;
+        auto inputs         = ins->inputs();
+        argument starts_arg = inputs.at(1)->eval();
+        argument ends_arg   = inputs.at(2)->eval();
+        argument axes_arg   = inputs.at(3)->eval();
+        if(not(starts_arg.empty() and ends_arg.empty() and axes_arg.empty()))
```
Comment (umangyadav) [inline src/simplify_dyn_ops.cpp:118]: same as previous comment. Need to apply `not` individually.

File: test/simplify_dyn_ops_test.cpp [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+            m1.add_instruction(migraphx::make_op("multibroadcast"), literal_ins, input);
+        auto add_ins = m1.add_instruction(migraphx::make_op("add"), input, broadcast_lit);
+        m1.add_return({add_ins});
+    }
+    run_pass(m1);
+
+    EXPECT(m0 == m1);
+}
+
+TEST_CASE(split_broadcast_match)
```
Comment (umangyadav) [inline test/simplify_dyn_ops_test.cpp:97]: I don't find `slice` in this test case.

File: test/simplify_dyn_ops_test.cpp [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+        auto input     = m0.add_parameter("data", s);
+        auto slice_ins = m0.add_instruction(
+            migraphx::make_op("slice", {{"starts", {0}}, {"ends", {3}}, {"axes", {0}}}), input);
+        m0.add_return({slice_ins});
+    }
+
+    migraphx::module m1;
+    {
+        migraphx::shape s{migraphx::shape::float_type, {6, 4, 4}};
+        auto input = m1.add_parameter("data", s);
```
Comment (umangyadav) [inline test/simplify_dyn_ops_test.cpp:199]: can you make input dynamic shaped for one of the tests or add another test ?

File: test/simplify_dyn_ops_test.cpp [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+        m0.add_return({add_ins});
+    }
+
+    migraphx::module m1;
+    {
+        migraphx::shape s{migraphx::shape::float_type, {2, 4}};
+        auto input = m1.add_parameter("data", s);
+        migraphx::shape lit_s{migraphx::shape{migraphx::shape::float_type, {4}}};
+        auto literal_ins = m1.add_literal(migraphx::literal{lit_s, {6, 5, 4, 3}});
+        auto broadcast_lit =
```
Comment (umangyadav) [inline test/simplify_dyn_ops_test.cpp:57]: I got confused a bit while reading. Generally, tests have input program/module defined first and then expected.

Here you have exptected defined first and then input.

File: test/simplify_dyn_ops_test.cpp (hunk 5 above)
Comment (CharlieL7) [reply test/simplify_dyn_ops_test.cpp:97]: That's intended, this is mimicking simplifying the broadcasts after a split_single_dyn_dim

File: test/simplify_dyn_ops_test.cpp (hunk 6 above)
Comment (CharlieL7) [reply test/simplify_dyn_ops_test.cpp:199]: Technically that's possible, but not intended with dynamic batch after split_single_dyn_dim. It should work though, will make the test.

File: test/simplify_dyn_ops_test.cpp (hunk 5 above)
Comment (umangyadav) [reply test/simplify_dyn_ops_test.cpp:97]: so `split` was to refer to the fact that input is output from `split_single_dyn_dim` pass ?

Comment (CharlieL7) [reply test/simplify_dyn_ops_test.cpp]: yes, I renamed the test to be more clear

**Review by umangyadav — APPROVED** [review]

**Pull request merged**